extension/contentfragment/bundle/src/main/java/com/adobe/cq/wcm/core/components/extension/contentfragment/internal/models/v1/ContentFragmentImpl.java

```diff
@@ -486,8 +485,45 @@ public String getHtml() {
                 return null;
             }
             // split into paragraphs
-            return content.split("(?=(<p>|<h1>|<h2>|<h3>|<h4>|<h5>|<h6>))");
+            return splitContentIntoParagraphs(content);
         }
+
+        private static String[] splitContentIntoParagraphs(@Nonnull String content) {
+            ArrayList<String> paragraphs = new ArrayList<>();
+
+            // wrap the content snippet into html (Jsoup)
+            Document contentDocument = Jsoup.parseBodyFragment(content);
+            Elements bodies = contentDocument.getElementsByTag("body");
+            // make sure it contains one only since it get wrapped normally
+            if (bodies.size() == 1) {
+                Node bodyNode = bodies.get(0); // access body element
+
+                // loop the root element under body and filter to only accept expected ones
+                for (Node childNode : bodyNode.childNodes()) {
+                    String nodeName = childNode.nodeName().toLowerCase();
+                    if (childNode instanceof org.jsoup.nodes.Element &&  StringUtils.containsAny(nodeName, "p", "h1", "h2", "h3", "h4", "h5")) {
+                        paragraphs.add(childNode.outerHtml());
+                    } else if (childNode instanceof TextNode) {
+                        TextNode txtNode = (TextNode) childNode;
+                        String text = txtNode.text();
+                        // filter whitespaces or carriage returns like characters found
+                        // in between the html elements
+                        if (!StringUtils.containsOnly(text,
+                                ' ',
+                                '\u00A0' /* nbsp */,
+                                '\t',
+                                '\n',
+                                '\r'
+                        )) {
+                            paragraphs.add(text);
```

Comment:
This should never happen, as there should be no `TextNode`s with actual content on this level - RTE should ensure that they are nested into container elements like `p`, `h?`, `li`, etc.

As RTE is not infallible, we can probably still keep it.

Reply:
It's mainly to be aligned with the core implementation where it also have this check in multiple places...